lease/lessor_bench_test.go

```diff
@@ -15,112 +15,237 @@
-func BenchmarkLessorRenew10(b *testing.B)      { benchmarkLessorRenew(10, b) }
-func BenchmarkLessorRenew100(b *testing.B)     { benchmarkLessorRenew(100, b) }
-func BenchmarkLessorRenew1000(b *testing.B)    { benchmarkLessorRenew(1000, b) }
-func BenchmarkLessorRenew10000(b *testing.B)   { benchmarkLessorRenew(10000, b) }
-func BenchmarkLessorRenew100000(b *testing.B)  { benchmarkLessorRenew(100000, b) }
-func BenchmarkLessorRenew1000000(b *testing.B) { benchmarkLessorRenew(1000000, b) }
-
-func BenchmarkLessorRevoke1(b *testing.B)       { benchmarkLessorRevoke(1, b) }
-func BenchmarkLessorRevoke10(b *testing.B)      { benchmarkLessorRevoke(10, b) }
-func BenchmarkLessorRevoke100(b *testing.B)     { benchmarkLessorRevoke(100, b) }
-func BenchmarkLessorRevoke1000(b *testing.B)    { benchmarkLessorRevoke(1000, b) }
-func BenchmarkLessorRevoke10000(b *testing.B)   { benchmarkLessorRevoke(10000, b) }
-func BenchmarkLessorRevoke100000(b *testing.B)  { benchmarkLessorRevoke(100000, b) }
-func BenchmarkLessorRevoke1000000(b *testing.B) { benchmarkLessorRevoke(1000000, b) }
+// NOTE: run with --benchtime Nx, please
 
-func benchmarkLessorFindExpired(size int, b *testing.B) {
-	lg := zap.NewNop()
-	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	le.Promote(0)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
+func BenchmarkLessorGrant10000(b *testing.B)  { benchmarkLessorGrant(10000, b) }
+func BenchmarkLessorGrant100000(b *testing.B) { benchmarkLessorGrant(100000, b) }
+
+func BenchmarkLessorRevoke10000(b *testing.B)  { benchmarkLessorRevoke(10000, b) }
+func BenchmarkLessorRevoke100000(b *testing.B) { benchmarkLessorRevoke(100000, b) }
+
+func BenchmarkLessorRenew10000(b *testing.B)  { benchmarkLessorRenew(10000, b) }
+func BenchmarkLessorRenew100000(b *testing.B) { benchmarkLessorRenew(100000, b) }
+
+func BenchmarkLessorFindExpired10000(b *testing.B)  { benchmarkLessorFindExpired(10000, b) }
+func BenchmarkLessorFindExpired100000(b *testing.B) { benchmarkLessorFindExpired(100000, b) }
+
+func benchmarkLessorGrant(size int, b *testing.B) {
+	// refresh after incr 50% lease
+	refresh := size / 2
+	if refresh < 1 {
+		refresh = 1
 	}
-	le.mu.Lock() //Stop the findExpiredLeases call in the runloop
-	defer le.mu.Unlock()
+
+	maxTTL := int64((size + refresh) / 100)
+	if maxTTL < 100 {
+		maxTTL = 100
+	}
+	ttls := make([]int64, size+refresh)
+	for i := 0; i < size+refresh; i++ {
+		ttls[i] = rand.Int63n(100) + 100
+	}
+
+	var count int
+	var le *lessor
+	var tearDown func()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		le.findExpiredLeases(1000)
+		if count > refresh || count == 0 {
+			count = 1
+			b.StopTimer()
+			if tearDown != nil {
+				tearDown()
+				tearDown = nil
+			}
+			le, tearDown = setUp()
+			for j := 0; j < size; j++ {
+				le.Grant(LeaseID(j+1), ttls[j])
```

Comment:
Do we need to hold `le.mu` in order to prevent leases from being auto-revoked by the lessor?

Reply:
auto-revoked means expire? We can limit the minimum ttl to avoid lease expire before test end.